pyro/contrib/epidemiology/compartmental.py

```diff
@@ -433,3 +456,45 @@ def enum_shape(position):
         logp = logp.reshape(-1).logsumexp(0)
         warn_if_nan(logp)
         pyro.factor("transition", logp)
+
+
+class _SMCModel:
+    """
+    Helper to initialize a CompartmentalModel to a feasible initial state.
+    """
+    def __init__(self, model):
+        assert isinstance(model, CompartmentalModel)
+        self.model = model
+
+    def init(self, state):
+        with poutine.trace() as tr:
+            params = self.model.global_model()
+        for name, site in tr.trace.nodes.items():
+            if site["type"] == "sample":
+                state[name] = site["value"]
+
+        self.t = 0
+        state.update(self.model.initialize(params))
+        self.step(state)  # Take one step since model.initialize is deterministic.
+
+    def step(self, state):
+        with poutine.block(), poutine.condition(data=state):
+            params = self.model.global_model()
```

Comment:
why is this needed exactly?

Reply:
Recall that `SMCFilter` tracks all state in a `state` dict and periodically resamples all that state. In this use case the `state` dict contains values of all sample sites that we will eventually replay. In particular it contains sample sites for global variables. Now since resampling changes those sites, and since `global_model()` not only draws samples but performs computation on those samples, we need to re-perform that computation after each resampling. Hence we block (to avoid this being traced again), and condition on the `SMCFilter`-managed `state` to get the latest resampled `params` each step.